Proposal for the new AppendVec storage

yhchiang-sol opened the pull request:
This PR includes a proposal that discusses three potential areas that we can further reduce
+the storage size of AppendVec:

* Use one byte for `executable` + `rent_epoch` fields (saves 8 bytes per account).
* Skip persisting the account hash for each account (saves ~32 bytes per account).
* Storing account data separately (better compression rate on account data).

jeffwashington commented:
goals:
1. improve compressability for snapshots
2. reduce how much has to be mapped into memory to load accounts
3. reduce file size to save disk, reduce snapshot size, reduce how much has to be read in/written and scanned/loaded/mapped
4. consider cpu/mem/disk tradeoffs.
5. eliminate redundant data

jeffwashington commented:
we've now activated the feature that stops updating rent epoch for rent exempt accounts.
we also don't allow creating rent paying accounts anymore
so, rent_epoch will soon become irrelevant.
rent_epoch is still a component of an accounts hash
at some point, we can stop storing/loading rent_epoch completely.
we should investigate the distrubution of rent_epoch and what we expect it to be in the future.
Should we do something to freeze rent_exempt accounts at a known rent_epoch (like SLOT::MAX) so that we can stop saving rent_epoch and use a bit to represent that the rent_epoch is fixed? I'll create a feature for this.

jeffwashington commented:
#28585
setting `rent_epoch` to known, constant value for rent exempt accounts.

jeffwashington commented:
Note that we are no longer updating rent_epoch with each epoch. As a result, we will now have a wide range of values for `rent_epoch` going forward. Previously they were all within a few epochs of the append vec's slot's epoch.

yhchiang-sol commented:
Thanks for sharing your thoughts, Jeff!  It really helps reorganize the unstructured ideas in my mind and boost more ideas.

Probably worth discussing the items/ideas one by one.

> * sorted by pubkey for some reason?

Sorting will provide two benefits:

* Enable binary search.  This will help reduce the account index size as we don't need to save the `reduced_offset` inside the `AccountInfo`, which is 4 bytes per account.  This enables us to trade CPU for smaller memory and storage usage for the accounts index.

* Possibly better compression rate as sorted pubkeys are more likely to share the same prefix, but this all depends on the density of the pubkeys.

yhchiang-sol commented:
> consider tradeoffs of 2 files per append vec vs 1 (storing data separately). Maybe we store them together, but in separate sections and maybe we compress the 2 halves separately for snapshots?

The very first version in my mind was one file, but I later found that if we want to keep supporting `append_ptr` using the existing mmapped append_vec file, then we are going to need two files as we are not able to "append" anymore since we have more than one section in the file.

I will keep thinking about what would be the best layout and format for AppendVec storage together with accounts index.

yhchiang-sol commented:
> figure out how to reduce the bytes use for # lamports. We already have unaligned data for 'data'. u64 is 8 bytes

For lamports, can I know what is the maximum allowed number of lamports per account?

> reduce the bytes used for data len. We already have unaligned data for 'data'

For this, I am thinking loud that if the basic unit of the data size is 1k or something, then our number here can just represent the size in terms of how many units.  (i.e., 1000 means 1000k).  In that case, u32 will be able to represent up to 4TB, which should be enough.

If the above idea sounds good, and if we are okay with 4 billion maximum for lamports, then we can use one single u64 to store both lamports and account data size?

yhchiang-sol commented:
> special case for 0 len data. basically, I'm imagining a byte or few bytes with a few common enum values (default account=0, account owned by system program with zero data=1, account owned by program '1' in lookup table with zero data=2, ...). Maybe it is a bit field, where 0 data is 1 bit. owner is 4 bits, lamports < 3 bytes is 1 bit, something like that.

Hmm, if we store all accounts data separately, then we can probably skip storing the account data size as we can derive the account data size by (the account data offset of the next account entry) - (the offset account data of the current account entry)

yhchiang-sol commented:
Updated the design.  The proposal now includes the general design and several graphs.
While I am still working on adding more text and details, the proposal should be ready for initial feedback. 

Here's a quick summary.  

The new AppendVec storage format uses ~80 bytes for each account on average,
and it's also more compression friendly for account data.

Specifically:

* Rent-exempt account w/o shared owner uses 80 bytes each.
* Rent-exempt account w/ shared owner uses 40 + 32 / N bytes each,
  where N is the number of accounts sharing its owner.
* Non-rent exempt account uses an extra 8 bytes for rent epoch.

The new AppendVec file is organized into 6 different blocks: header,
account metas block, account pubkeys block, owners block, account data block,
and footer.

```
+---------------------------------------------------+
| header                | 40 bytes                  |
+---------------------------------------------------+
| account metas block   | 16 bytes per account      |
+---------------------------------------------------+
| account pubkeys block | 32 bytes per account      |
+---------------------------------------------------+
| owners block          | 32 bytes per unique owner |
+---------------------------------------------------+
| account data block    | N + optional 8 bytes      |
|                       | for rent epoch            |
+---------------------------------------------------+
| footer                | 40 bytes                  |
+---------------------------------------------------+

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+    },
+    pub account_meta: &'a AccountMeta {
+        pub lamports: u64,     //  8 bytes
+        pub owner: Pubkey,     // 32 bytes, encoded from a 44-byte string
```
This is not encoded from a string. This is the representation (Pubkey). It can be displayed as a string. And, the Pubkey can be transformed to/from a string.

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+
+## Problems
+The storage format, StoredAccountMeta, of each entry in the AppendVec currently
+has 129 bytes overhead + the size of account data for each account.
```
plus it is aligned

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+### Per-Account Hash is replaced by AppendVec Block Hash
+The per-account hash is removed.  Instead, each block in the new AppendVec file
+has one 32-bytes hash, and the hash for the entire AppendVec is also included
+in the footer.
```
why a hash in the footer for this append vec? This is not deterministic per machine.

yhchiang-sol replied:
The footer includes the AppendVec hash + a magic number.  If the ending part of one AppendVec files does not match the predefined magic number, then we know for sure that the file has been truncated or incompleted.

yhchiang-sol replied:
Removed the footer.  Move the AppendVec hash to the header instead.

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+| Account Meta (16 bytes each)                                |
++-------------------------------------------------------------+
+| owner local id  | 4 bytes                                   |
+|                 | up to 4,294,967,295 owners per append vec |
```
currently, max file size is 16G:
https://github.com/solana-labs/solana/blob/37507a2de636efec8bec0b38e794e22230c8d183/runtime/src/append_vec.rs#L42
That could inform our max ranges.

yhchiang-sol replied:
> currently, max file size is 16G:

Good to know this.

The local ID represents the ith entry in the owner block.  If the max file size is 16G, then the number of bytes used for the owner's local id can be smaller. (basically, the maximum number of owners allowed in one AppendVec.)

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+|                 | up to 4,294,967,295 owners per append vec |
+| data_offset     | 4 bytes                                   |
++-----------------+-------------------------------------------+
+| rent status     | 1 bit                                     |
```
this is able to be deterministically calculated based on lamports and data len

yhchiang-sol replied:
Thanks for pointing this out.  Yes!  So we can remove this bit!

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+                       // the id can be used to fetch its pubkey
+    pub owner_local_id: u32,
+    data_offset: u32,
+    data_size: u32,   // derived when first loading the AppendVec from the storage
```
fwiw, data size is currently limited to 10MB, iirc.

yhchiang-sol replied:
Do we happen to have some size unit for the data size?  Such as 1k or 64 bytes or anything like this?  If we have a size unit, then we can represent a wider range of data sizes with fewer bytes.

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+| Owner (40 bytes each)                                       |
++-------------------------------------------------------------+
+| pubkey          | 32 bytes                                  |
+| write version   |  8 bytes                                  |
```
We can get rid of write version if we know there is only 1 append vec per slot. These are all theoretical comments, not all the way thought through.

yhchiang-sol replied:
> We can get rid of write version if we know there is only 1 append vec per slot.
Good to know this.  Looks like we can save 8 bytes per owner at least for now.

Btw, as you mentioned earlier that one AppendVec is currently capped at 16GB, can I know what will happen if there's a hot slot such that it goes beyond the 16GB limit?

yhchiang-sol replied:
Offline discussed with Jeff.  Supporting multiple AppendVec per slot to enable multiple writers is not currently needed as we already have a write cache on top of AppendVec.

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
++------------------------------------+
+| Data blocks hash | 32 bytes        |
++------------------------------------+
+| Account data 1                     |
```
I think I like the removal of padding here.

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
++------------------------------------+
+| AppendVec Hash | 32 bytes          |
++------------------------------------+
+| Magic number   |  8 bytes          |
```
This footer (and the format) will make these append vecs not usable for the ancient append vec process at the moment.

yhchiang-sol replied:
Thanks for pointing this out.  Let me think more about it for ancient AppendVecs.

jeffwashington commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
++-------------------------------------------------------------+
+| Account data block (variable size)                          |
++-------------------------------------------------------------+
+| rent_epoch      |  0-8 bytes                                |
```
I think this may have to be aligned.

yhchiang-sol replied:
Feel the same way.  Then we might need to store the account data size for each account as there might be gaps between two account data entries with paddings.

alessandrod replied:
How is this accessed? I imagine it's read, not directly mmapped, so it can be left unaligned and read with https://doc.rust-lang.org/stable/std/ptr/fn.read_unaligned.html

yhchiang-sol commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,231 @@
+
+### Per-Account Hash is replaced by AppendVec Block Hash
+The per-account hash is removed.  Instead, each block in the new AppendVec file
+has one 32-bytes hash, and the hash for the entire AppendVec is also included
```
IMO one hash per AppendVec should be sufficient, but if we want to know which part goes wrong, then we can do one hash per block.  Note that one block means one type of data.  So probably section might be a better word.

yhchiang-sol replied:
Updated the proposal.  The footer and the per-block hash are currently removed from the design.  Now it uses one hash for the entire AppendVec + some sanity check for the size and offset of each block.

yhchiang-sol reviewed (commented):
Thanks for the initial feedback, @jeffwashington!  Will address the alignment for the data block and extend the design to cover ancient AppendVec.

yhchiang-sol commented:
Updated the proposal:
* Removed write_version.
* Added account meta entry size and owner entry size in the header to enable forward compatibility.
* Added alignment information.
* More descriptions.

yhchiang-sol commented:
Updated the proposal to include compression and the concept of the account data block.

t-nelson commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,374 @@
++-------------------------------------------------------------+
+| Account data (after decompression)                          |
++-------------------------------------------------------------+
+| rent_epoch      |  0-8 bytes                                |
```
zero _or_ 8 bytes, right?  this should probably be appended instead of prepended and `NonZeroU64`, then we get free memory layout optimization from the compiler (assuming rent-epoch == 0 is an invalid thing)

yhchiang-sol replied:
Thanks for the correction.  Yes, it should be 0 or 8 bytes.

If we read account data more often than rent-epoch (I think this assumption should be yes), then yes it should be appended instead of prepended.

jeffwashington replied:
`rent_epoch` is quite the story. It is only relevant anymore for rent-paying accounts, which is a shrinking set of ~1M accounts on mnb. For everyone else, I have a feature to set the value to Epoch::MAX for all rent exempt accounts. Thus, we don't need to store it at all for rent exempt accounts (the vast bulk of accounts). For rent paying accounts it is 1 of 4 values. Slot - 1, Slot, Slot + 1, 0. Where Slot is the slot of the append vec itself. It is possible we could reduce these values down to fewer than 4. Maybe even 2 values:
1. rent already collected as of this slot
2. rent last collected as of prior slot

jeffwashington replied:
I am hustling to get this feature ready.
https://github.com/solana-labs/solana/issues/28683

t-nelson commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,374 @@
+and does not change the definition of the existing fields.
+
+#### (Small) Account Data Blocks
+One small account data block contains multiple accounts' data.
```
is each data entry a full page, or variable-length?

cross-page load/store perf is usually pretty abysmal.  like an order of magnitude bad

t-nelson reviewed (commented):
just some first pass thoughts for now.  consider getting input from the firedancer team as well.  they are highly skilled in data path optimization and will need to work with us at least wrt snapshot format

t-nelson commented:
This is a good read if we're considering retaining mmap-based file-backing. tl;dr, we probably shouldn't

[Are You Sure You Want to Use MMAP in Your Database Management System?](https://www.cidrdb.org/cidr2022/papers/p13-crotty.pdf) (pdf)

yhchiang-sol commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,374 @@
+offers forward compability as long as the newer version only adds new fields
+and does not change the definition of the existing fields.
+
+#### (Small) Account Data Blocks
```
The current doc here is a bit outdated as I am currently prototyping.

In the current prototype, the account data blocks will be the first thing in the file, so the first block is guaranteed to be page-aligned. But since each block is compressed, the rest of the blocks will not be page-aligned.

Let me think if we could get benefits from both compress and page alignment.

yhchiang-sol commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,374 @@
+| data_block_offset            |  8 bytes     |
+| blob_data_block_offset       |  8 bytes     |
+| compression_algorithm        |  8 bytes     |
+| append_vec_hash              | 32 bytes     |
```
Great we share the same idea.  The current prototype uses the footer approach instead.  Let me update the doc later today.

yhchiang-sol commented:
Thanks for the feedback, @t-nelson!  The doc is currently out-of-date as I am implementing the prototype but many of your comments are still applied!  Will update the doc later today and have fire-dancers folks involved.

> This is a good read if we're considering retaining mmap-based file-backing. tl;dr, we probably shouldn't
> [Are You Sure You Want to Use MMAP in Your Database Management System?]
> (https://www.cidrdb.org/cidr2022/papers/p13-crotty.pdf) (pdf)

Rest assured that the new design will not use MMAP :). We will LRU or some other suitable cache mechanisms for reads.

yhchiang-sol commented:
Am still fixing bugs in the prototype.  I want to make sure everything runs well before updating the proposal.  Converting this PR to draft.

t-nelson commented:
fwiw, typically a proposal should be accepted before the implementation is started :wink:

yhchiang-sol commented:
> fwiw, typically a proposal should be accepted before the implementation is started 

Yep yep, it's just a prototype to make sure it can actually produce the expected outcome.

And here it is: https://github.com/solana-labs/solana/pull/28790

So far I am seeing up to ~75% size reduction, which is good!  Instructions to test the file format is also under the comment in #28790.

Will update the proposal shortly.

yhchiang-sol commented:
Hello @jeffwashington and @t-nelson,

Thanks for the feedback so far!

While I am still trying to add more sections/descriptions and polish the document, I've included the most important sections and information that allow our discussion to proceed.

Let me know what you think about the new format.

I've prototyped the new format and tested it by converting ~105GB append-vec files.  The new format only uses ~40GB.
If I further compress the files, it is ~30GB, which can be used as an estimation for its snapshot size.

yhchiang-sol commented on docs/src/proposals/append-vec-storage.md:
```diff
@@ -0,0 +1,374 @@
+title: Reduce the size of AppendVec storage in AccountsDB
+---
+
+## Problems
```
Am working on it!
